pkg/kubelet/kubelet_volumes.go

```diff
@@ -152,9 +153,14 @@ func (kl *Kubelet) removeOrphanedPodVolumeDirs(uid types.UID) []error {
 	}
 
 	// Remove any remaining subdirectories along with the volumes directory itself.
-	// Fail if any regular files are encountered.
+	// Fail if any regular files other than CSI volume data file are encountered.
 	podVolDir := kl.getPodVolumesDir(uid)
-	if err := removeall.RemoveDirsOneFilesystem(kl.mounter, podVolDir); err != nil {
+	if err := removeall.RemoveAllOneFilesystemCommon(kl.mounter, podVolDir, func(path string) error {
```

Comment:
This change actually re-introduces a rather serious race condition where `vol_data.json` may be deleted by `removeOrphanedPodVolumeDirs` before `NodeUnstageVolume` has a chance to run, which causes unmount to fail and kubelet ends up not being able to clean the mountpoints on the node:
https://github.com/kubernetes/kubernetes/issues/101911
So this is not the right place to remove `vol_data.json`, it will result in unmounts failing in some scenarios.
You may want to take a look at the [volume reconstruction](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/volumemanager/reconciler/reconciler.go) code in volumemanager. This is a tough bug in a complex part of the code, and the fix is not obvious to me, but when I saw it before [syncStates failed here](https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/volumemanager/reconciler/reconciler.go#L506) because the /mount subdir did not exist. Then it calls `cleanupMounts` but that failed because the volume failed to be reconstructed and it couldn't create the unmounter at that point. If you're able to reproduce the problem, you can check if you're seeing the same failure scenario in kubelet.log.
IMO, we either need to be able to reconstruct the volume to unmount it in this scenario, or at least find some way to create the unmounter so that cleanupMounts can remove vol_data.json even when the /mount subdir does not exist. Sorry if this direction is still a little vague, but that's as far as I've been able to dig into this.

Reply:
@dobsonj Thank you so much for the explanation. I am excited that there are still so many things I didn't know😄, and I will dig into it again to see if I can find a correct solution.

Reply:
@dobsonj  @gnufied  Actually, not only the volume directory `/mount` will be left over, the `/globalmount` directory will be left over either. Because in this scenario, the volume manager did neither of `NodeUnstageVolume` nor `NodeUnpublishVolume`:

When reconstructing, the volume directory is not mountpoint, so volume manager tries to do cleanup by calling `NodeUnpublishVolume` and cleanup the `/mount` directory and `vol_data.json`, but the csi driver did not register yet, which makes cleanup fails. Actually, no matter it succeeds or fails, `NodeUnstageVolume` won't be called, so `/globalmount` will be left over.

I think there may be two solutions:
1. Let the reconstruction success no matter `/mount` is mountpoint or not. After adding the volume to ASW, the volume manager's reconcile process will do `NodeUnstageVolume` and `NodeUnpublishVolume` later. 
2. When the unmounter executes unmounting, clean up the directory and `vol_data.json` without calling `NodeUnpublishVolume` and `NodeUnstageVolume` if the directory is already not a mountpoint.

WDYT? Which one is better or any other idea?